src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceParameterService.cs

```diff
@@ -0,0 +1,862 @@
+            CancellationToken cancellationToken)
+        {
+            var generator = SyntaxGenerator.GetGenerator(currentDocument);
+            var semanticFacts = currentDocument.GetRequiredLanguageService<ISemanticFactsService>();
+            var invocationSemanticModel = await currentDocument.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+
+            var root = await currentDocument.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var editor = new SyntaxEditor(root, generator);
+            var syntaxFacts = currentDocument.GetRequiredLanguageService<ISyntaxFactsService>();
+            var insertionIndex = GetInsertionIndex(compilation, methodSymbol, syntaxFacts, containingMethod);
+            var newMethodIdentifier = methodSymbol.Name + "_" + parameterName;
+            var validParameters = methodSymbol.Parameters.AsEnumerable().SelectMany(parameter => mappingDictionary.Values.Where(param => parameter.Equals(param))).ToImmutableArray();
+            if (trampoline)
+            {
+                foreach (var invocationExpression in invocations)
+                {
+                    editor.ReplaceNode(invocationExpression, (currentInvocation, _) =>
+                    {
+                        return GenerateNewInvocationExpressionForTrampoline(syntaxFacts, semanticFacts, invocationSemanticModel, generator, currentInvocation,
+                            invocationExpression, validParameters, newMethodIdentifier, insertionIndex, cancellationToken);
+                    });
+                }
+            }
+
+            // If you are at the original document, then also introduce the new method and introduce the parameter.
+            if (currentDocument.Id == originalDocument.Id)
+            {
+                if (trampoline)
+                {
+                    var newMethodNode = await ExtractMethodAsync(originalDocument, expression, methodSymbol, validParameters, newMethodIdentifier,
+                        generator, cancellationToken).ConfigureAwait(false);
+                    editor.InsertBefore(containingMethod, newMethodNode);
+                }
+                if (overload)
+                {
+                    var newMethodNode = await GenerateNewMethodOverloadAsync(originalDocument, expression, methodSymbol, insertionIndex,
+                        generator, cancellationToken).ConfigureAwait(false);
+                    editor.InsertBefore(containingMethod, newMethodNode);
+                }
+
+                await UpdateExpressionInOriginalFunctionAsync(originalDocument, expression, containingMethod,
+                    parameterName, editor, allOccurrences, cancellationToken).ConfigureAwait(false);
+                var parameterType = invocationSemanticModel.GetTypeInfo(expression, cancellationToken).ConvertedType ?? invocationSemanticModel.Compilation.ObjectType;
+                var refKind = syntaxFacts.GetRefKindOfArgument(expression);
```

Comment:
this line doesn't make sense.  it consumes an argument, but you're passing an expression.  i don't believ ethis could do anything.  is thtere a test that demosntrates why this line is needed.

Reply:
ah right, I meant to remove this. It does not do anything and ended up defaulting to RefKind.None in any case.